File: flutter/craftd_widget/lib/presentation/ui/text/craftd_text_builder.dart
import 'package:flutter/src/widgets/framework.dart';
import 'package:craftd_widget/data/model/text/text_properties.dart';

import '../../../data/model/action/action_properties.dart';
import '../../builder/craftd_builder.dart';
import '../../craftd_view_listener.dart';
import 'craftd_text.dart';

class CraftDTextBuilder extends CraftDBuilder<TextProperties> {
  CraftDTextBuilder() : super(key: keyBuilder);
  static String keyBuilder = 'CraftDTextView';

  @override
  Widget craft(TextProperties model, CraftDViewListener listener) {
    return CraftDText(
        textProperties: model,
        callback: () {
          if (model.actionProperties != null) {
            listener(model.actionProperties!);
          }
        });
  }

  @override
  TextProperties fromJson(properties) {
    return TextProperties(
        text: properties['text'],
        textColorHex: properties['textColorHex'],
        textSize: properties['textSize'],
        textAllCaps: properties['textAllCaps'],
        backgroundHex: properties['backgroundHex'],
        actionProperties: ActionProperties.fromJson(
            properties[ActionProperties.key]));
  }

}

File: android_kmp/craftd-compose/src/commonMain/kotlin/com/github/codandotv/craftd/compose/ui/text/CraftDTextBuilder.kt
package com.github.codandotv.craftd.compose.ui.text

import androidx.compose.runtime.Composable
import com.github.codandotv.craftd.androidcore.data.convertToElement
import com.github.codandotv.craftd.androidcore.data.model.base.SimpleProperties
import com.github.codandotv.craftd.androidcore.data.model.text.TextProperties
import com.github.codandotv.craftd.androidcore.presentation.CraftDComponentKey
import com.github.codandotv.craftd.androidcore.presentation.CraftDViewListener
import com.github.codandotv.craftd.compose.builder.CraftDBuilder

class CraftDTextBuilder(
    override val key: String =  CraftDComponentKey.TEXT_VIEW_COMPONENT.key
) : CraftDBuilder{
    @Composable
    override fun craft(model: SimpleProperties, listener: CraftDViewListener) {
        val textProperties = model.value.convertToElement<TextProperties>()
        textProperties?.let {
            CraftDText(textProperties = it) {
                textProperties.actionProperties?.let { listener.invoke(it) }
            }
        }
    }
}